contracts/EntryPoint.sol

```diff
@@ -191,30 +185,35 @@ contract EntryPoint is StakeManager {
     function simulateValidation(UserOperation calldata userOp) external returns (uint256 preOpGas, uint256 prefund) {
         uint256 preGas = gasleft();
 
-        bytes32 requestId = getRequestId(userOp);
-        (prefund,,) = _validatePrepayment(0, userOp, requestId);
+        UserOpInfo[] memory opInfos = new UserOpInfo[](1);
+        _validatePrepayment(0, userOp, opInfos);
+        prefund = opInfos[0].prefund;
         preOpGas = preGas - gasleft() + userOp.preVerificationGas;
 
         require(msg.sender == address(0), "must be called off-chain with from=zero-addr");
     }
 
-    function _getPaymentInfo(UserOperation calldata userOp) internal view returns (uint256 requiredPrefund, PaymentMode paymentMode) {
-        requiredPrefund = userOp.requiredPreFund();
-        if (userOp.hasPaymaster()) {
-            paymentMode = PaymentMode.paymasterDeposit;
-        } else {
-            paymentMode = PaymentMode.walletDeposit;
-        }
+    function _getPaymentInfo(UserOperation calldata userOp) internal view returns (uint256 requiredPrefund, address paymaster) {
+    unchecked {
+        paymaster = userOp.paymaster;
```

Comment:
There seems to be no point in returning `address paymaster` unmodified here.

Reply:
I use the return value to cache into the opInfo